File: 1091-shortest-path-in-binary-matrix/1091-shortest-path-in-binary-matrix.js
/**
 * @param {number[][]} grid
 * @return {number}
 */
const shortestPathBinaryMatrix = (grid) =>{
    if(grid[0][0] ===1) {
        return -1
    } 
    const queue = [[0, 0, 1]]
    let head = 0
    grid[0][0] = 1
    const n= grid.length
    if(n===1) return 1


    const dirs = [[-1,-1],[-1,0],[-1,1],[0,-1],[0,1],[1,-1],[1,0],[1,1]]
    while(head<queue.length){
        const [r,c,dist] = queue[head++]
        for(const [dr,dc] of dirs){
            const nr = r + dr
            const nc = c + dc

            if(nr<0 || nc<0 || nr>=n || nc>=n) continue
            if(grid[nr][nc] === 1 ) continue
            if(nr === n-1 && nc === n-1) return dist+1

            grid[nr][nc] = 1
            queue.push([nr,nc,dist+1])
        }
    }
    return -1
};

File: 1091-shortest-path-in-binary-matrix/1091-shortest-path-in-binary-matrix.py
from collections import deque

class Solution:
    def shortestPathBinaryMatrix(self, grid: List[List[int]]) -> int:
        row = len(grid)
        col = len(grid[0])
        queue = deque([(0,0,1)])

        dr = [-1,1,0,0,-1,-1,1,1] #상하좌우 위위아래아래
        dc = [0,0,-1,1,-1,1,-1,1]

        if (grid[0][0] ==1 or grid[row-1][col-1] == 1):
            return -1
        
        
        while queue:
            r,c,d= queue.popleft()
            if(r == row-1 and c == col-1):
                return d
            for i in range(8):
                nr = r+dr[i]
                nc = c+dc[i]

                
                if(nr >=0 and nc >= 0 and nr <row and nc< col and grid[nr][nc] == 0):
                    queue.append((nr,nc,d+1))
                    grid[nr][nc] = 1
        
        return -1




            
            

        
        